src/braket/experimental/autoqasm/api.py

```diff
@@ -320,6 +372,60 @@ def _convert_program_as_subroutine(
         root_oqpy_program._add_subroutine(subroutine_name, subroutine_function_call.subroutine_decl)
 
 
+def _convert_gate(
+    f: Callable,
+    declaration_only: bool,
+    options: converter.ConversionOptions,
+    args: List[Any],
+    kwargs: Dict[str, Any],
+) -> Callable:
+    try:
+        # We must be inside an active conversion context in order to invoke a gate
+        program_conversion_context = aq_program.get_program_conversion_context()
+        oqpy_program = program_conversion_context.get_oqpy_program()
+
+        # Wrap the function into an oqpy gate definition
+        wrapped_f, gate_args = _wrap_for_oqpy_gate(f, options)
+        gate_name = f.__name__
+        qubit_names = [name for name, is_qubit in gate_args if is_qubit]
+        qubits = [oqpy.Qubit(name, needs_declaration=False) for name in qubit_names]
+        if not qubits:
+            raise errors.ParameterTypeError(
+                f'Gate definition "{gate_name}" has no arguments of type aq.Qubit. '
+                "Every gate definition must contain at least one qubit argument."
+            )
+
+        if not declaration_only:
+            # Process the gate definition
+            angle_names = [name for name, is_qubit in gate_args if not is_qubit]
+            angles = [oqpy.AngleVar(name=name) for name in angle_names]
+            with program_conversion_context.gate_definition(gate_name, qubits, angles):
+                # TODO - enforce that nothing gets added to the program inside here except gates
+                wrapped_f(qubits, *angles)
+
+            # Add the gate definition to the root-level program if necessary
+            root_oqpy_program = program_conversion_context.oqpy_program_stack[0]
+            if gate_name not in root_oqpy_program.gates:
+                root_oqpy_program._add_gate(gate_name, oqpy_program.gates[gate_name])
+
+        # Add the gate invocation to the program
+        if len(args) != len(gate_args):
+            raise errors.ParameterTypeError(
+                f'Incorrect number of arguments passed to gate "{gate_name}". '
+                f"Expected {len(gate_args)}, got {len(args)}."
+            )
+        qubit_args = [args[i] for i, (_, is_qubit) in enumerate(gate_args) if is_qubit]
+        angle_args = [args[i] for i, (_, is_qubit) in enumerate(gate_args) if not is_qubit]
+        aq_instructions.instructions._qubit_instruction(gate_name, qubit_args, *angle_args)
```

Comment:
The argument handling here may be somewhat confusing and I'm open to suggestions. I wanted to allow the Python function signature to be flexible and allow qubit and angle parameters in any arbitrary order. But because OpenQASM treats qubit and angle argument lists separately, the Python argument list has to be separated in order to add the gate to the oqpy program.

Reply:
It feels a bit confusing to me, but I think it's just that it's lower-level than most of the rest of the function. Could this be factored out into a function for readability?